app/src/main/scala/org/alephium/explorer/persistence/queries/BlockQueries.scala

```diff
@@ -213,4 +219,87 @@ trait BlockQueries
       .result
       .headOption
   }
+
+  /** Inserts block_headers or update them if there is a primary key conflict */
+  // scalastyle:off method.length magic.number
+  def upsertBlockHeaders(blocks: Iterable[BlockHeader]): DBActionW[Int] =
+    if (blocks.isEmpty) {
+      DBIOAction.successful(0)
+    } else {
+      val placeholder = paramPlaceholder(rows = blocks.size, columns = 14)
+
+      val query =
+        s"""
+           |insert into $block_headers ("hash",
+           |                            "timestamp",
+           |                            "chain_from",
+           |                            "chain_to",
+           |                            "height",
+           |                            "main_chain",
+           |                            "nonce",
+           |                            "version",
+           |                            "dep_state_hash",
+           |                            "txs_hash",
+           |                            "txs_count",
+           |                            "target",
+           |                            "hashrate",
+           |                            "parent")
+           |values $placeholder
+           |ON CONFLICT ON CONSTRAINT block_headers_pkey
+           |    DO NOTHING
+           |""".stripMargin
+
+      val parameters: SetParameter[Unit] =
+        (_: Unit, params: PositionedParameters) =>
+          blocks foreach { block =>
+            params >> block.hash
+            params >> block.timestamp
+            params >> block.chainFrom
+            params >> block.chainTo
+            params >> block.height
+            params >> block.mainChain
+            params >> block.nonce
+            params >> block.version
+            params >> block.depStateHash
+            params >> block.txsHash
+            params >> block.txsCount
+            params >> block.target
+            params >> block.hashrate
+            params >> block.parent
+        }
+
+      SQLActionBuilder(
+        queryParts = query,
+        unitPConv  = parameters
+      ).asUpdate
+    }
+  // scalastyle:on method.length magic.number
+
+  /** Transactionally write blocks */
+  @SuppressWarnings(Array("org.wartremover.warts.MutableDataStructures"))
+  def upsertBlockEntity(blocks: Iterable[BlockEntity], groupNum: Int): DBActionRWT[Int] = {
+    val blockDeps    = ListBuffer.empty[BlockDepEntity]
+    val transactions = ListBuffer.empty[TransactionEntity]
+    val inputs       = ListBuffer.empty[InputEntity]
+    val outputs      = ListBuffer.empty[OutputEntity]
+    val blockHeaders = ListBuffer.empty[BlockHeader]
```

Comment:
Is `ListBuffer` more performant than `ArrayBuffer`? I use `Array` based collections, instead of `List` based collections in general

Reply:
`ListBuffer` in this case, appends are constant time.  I've never used Array based collections or `ArrayBuffer` as it copies its internal array on overflow to fit new data. If I know the collection size before hand I'd use `Array` directly. Please see [performance-characteristics](https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html) for details.